src/Integration/Notifications/SonarQubeNotifications.cs

```diff
@@ -0,0 +1,141 @@
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.ComponentModel.Composition;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Threading;
+using System.Timers;
+using Microsoft.VisualStudio.Shell;
+using SonarLint.VisualStudio.Integration.Service;
+using SonarLint.VisualStudio.Integration.State;
+using SystemInterface.Timers;
+using SystemWrapper.Timers;
+
+namespace SonarLint.VisualStudio.Integration.Notifications
+{
+    [Export(typeof(ISonarQubeNotifications))]
+    internal class SonarQubeNotifications : ISonarQubeNotifications
+    {
+        public event EventHandler ShowDetails;
+
+        private INotifyIcon notifyIcon;
+        private readonly ITimer timer;
+        private const string iconPath = "pack://application:,,,/SonarLint;component/Resources/sonarqube_green.ico";
+        private const string tooltipTitle = "SonarQube notification";
+        private string message;
+        private DateTimeOffset lastRequestDate = DateTimeOffset.MinValue;
+        private readonly ISonarQubeServiceWrapper sonarQubeService;
+        private readonly IStateManager stateManager;
+        private readonly CancellationTokenSource cancellation = new CancellationTokenSource();
+
+        [ImportingConstructor]
+        [ExcludeFromCodeCoverage] // Do not unit test MEF constructor
+        internal SonarQubeNotifications(IHost host)
+            : this(host.SonarQubeService, host.VisualStateManager, new StatusBarIconWrapper(),
+                  new TimerWrap())
+        {
+        }
+
+        internal SonarQubeNotifications(ISonarQubeServiceWrapper sonarQubeService,
+            IStateManager stateManager, INotifyIcon notifyIcon, ITimer timer,
+            double timerIntervalMilliseconds = 60000 /* 60sec */)
+        {
+            this.notifyIcon = notifyIcon;
+            this.timer = timer;
+            this.sonarQubeService = sonarQubeService;
+            this.stateManager = stateManager;
+
+            timer.Elapsed += OnTimerElapsed;
+            timer.Interval = timerIntervalMilliseconds;
+        }
+
+        public void Start()
+        {
+            notifyIcon.BalloonTipClick += OnBalloonTipClick;
+            notifyIcon.IsVisible = true;
+
+            var serverConnection = ThreadHelper.Generic.Invoke(() => stateManager.GetConnectedServers().FirstOrDefault());
```

Comment:
_serverConnection_ isn't being used. However, since it looks like _Start_ is called on every bound solution changed event, it could be stored in a field. That would avoid the refetching the server connection each time the timer elapses and the message is updated.  Same goes for the project id.

Reply:
Very good point - fetching only in Start() makes sense.